volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+            ),
+            requirements.StringRequirement(
+                name="symbol",
+                description="Symbol name to resolve",
+                optional=True,
+            ),
+            requirements.IntRequirement(
+                name="address",
+                description="Address of symbol to resolve",
+                optional=True,
+            ),
+        ]
+
+    @staticmethod
+    def _get_pefile_obj(
+        context: interfaces.context.ContextInterface,
+        pe_table_name: str,
+        layer_name: str,
+        base_address: int,
+    ) -> Optional[pefile.PE]:
+        """
+        Attempts to pefile object from the bytes of the PE file
+
+        Args:
+            pe_table_name: name of the pe types table
+            layer_name: name of the process layer
+            base_address: base address of the module
+
+        Returns:
+            the constructed pefile object
+        """
+        pe_data = io.BytesIO()
+
+        try:
+            dos_header = context.object(
+                pe_table_name + constants.BANG + "_IMAGE_DOS_HEADER",
+                offset=base_address,
+                layer_name=layer_name,
+            )
+
+            for offset, data in dos_header.reconstruct():
+                pe_data.seek(offset)
+                pe_data.write(data)
+
+            pe_ret = pefile.PE(data=pe_data.getvalue(), fast_load=True)
+
+        except exceptions.InvalidAddressException:
+            pe_ret = None
+
+        return pe_ret
+
+    @staticmethod
+    def range_info_for_address(
+        ranges: ranges_type, address: int
+    ) -> Optional[range_type]:
+        """
+        Helper for getting the range information for an address
+        """
+        for start, size, filepath in ranges:
+            if start <= address < start + size:
+                return start, size, filepath
+
+        return None
+
+    @staticmethod
+    def filepath_for_address(ranges: ranges_type, address: int) -> Optional[str]:
+        """
+        Helper to get the file path for an address
+        """
+        info = PESymbols.range_info_for_address(ranges, address)
+        if info:
+            return info[2]
+
+        return None
+
+    @staticmethod
+    def filename_for_path(filepath: str) -> str:
+        """
+        Consistent way to get the filename
+        """
+        return filepath.split("\\")[-1]
+
+    @staticmethod
+    def addresses_for_process_symbols(
+        context: interfaces.context.ContextInterface,
+        config_path: str,
+        layer_name: str,
+        symbol_table_name: str,
+        symbols: filter_modules_type,
+    ) -> found_symbols_type:
+        collected_modules = PESymbols.get_process_modules(
+            context, layer_name, symbol_table_name, symbols
+        )
+
+        found_symbols = PESymbols.find_symbols(
+            context, config_path, symbols, collected_modules
+        )
+
+        for mod_name, unresolved_symbols in symbols.items():
```

Comment:
I can change the name of the variable, but modifying in the run is the cleanest to ensure we break as sound as everything is resolved.

Reply:
Are we likely to break during the run?  What would happen after that, how would we continue (since the other values haven't been tested)?  Surely it'd be better to catch breaks in the loop and then return at the end with as complete a set of answers as possible?  Expecting a break we haven't caught and using that as a reason to modify an input parameter seems like using a coding issue to justify making another coding issue?

Otherwise we could returning a failure case rather than modify the variable during run?

Reply:
The idea is to make it where the caller gets the symbols resolved and the leftover ones can be reported plus also not iterating over processes/pdbs/export tables once we have already resolved the requested symbols.

Reply:
Much better, thanks!

Reply:
I might've been too hasty, the symbols (input parameter) is still modified by certain functions in the above call?  Did you point out where the breaking could happen in that?  We have to complete the function to get an output for found_symbols, so even if we break (which should be caught and gracefully dealt with), we're still gonna have to wait for the final list, so why not wait for the final list *and* the unresolved symbols as a tuple?  I think (and hope) this is the *only* place in the code where we modify an input parameter, and I feel like we're suddenly changing the rules on people expecting things to work a particular way.

If you could spell out the specific scenario where having partial results in symbols is valuable, rather than returning a new list containing the resolved values (which the caller can then use to overwrite symbols explicitly) then I'll consider it, but at the moment it's just a very different style of coding from anything in the codebase, and I can't see a good reason for having it?

Reply:
Look at how unhooked_system_calls uses it. It first makes its data structure for the symbols it wants in ntdll.dll:

https://github.com/volatilityfoundation/volatility3/blob/b40c20dd7fe0d1e348f94531c290a9aceee34dcb/volatility3/framework/plugins/windows/unhooked_system_calls.py#L22

It then makes one simple function call to resolve al those symbols to their (name, address) pairs:

https://github.com/volatilityfoundation/volatility3/blob/b40c20dd7fe0d1e348f94531c290a9aceee34dcb/volatility3/framework/plugins/windows/unhooked_system_calls.py#L146

Along the way, the `system_calls` variable it send is in modified, but the caller doesn't care as its never going to use it again. The form being sent in isn't even useful after as what the caller cares about is is the (name, address) pairs that were resolved. We could make a copy of the sent in parameter if you really want it that way but even that is a waste of space and time as the caller is never going to reuse the values it sent in.

The input one is modified so we can break processing once all the symbols are resolved. So for example, if a process early in the prcoess list has the PDB for the ntdll.dll module, then all the symbols should resolve out of that. We detect that as the symbols are removed from the input list as they are found. We then break all processing (and preventing further PDB searches and export table reconstruction in later processes in the list) once all are done. The sme thing applies for kernel based symbols.

The current API is *extremely* clean to callers - its one function call to resolve any combination of modules and symbol names and/or symbol addresses. Having callers do anything but one function call is just putting extra effort and repeat code on callers, and I really can't see why we would want to do that and repeat code is gross anyway.